### PR TITLE
Validation - incident fields exist in layout and mappers

### PR DESCRIPTION
**Pull request by ChanochShayner**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/29841
fixes: https://github.com/demisto/etc/issues/31851

## Description
Validation - incident fields exist in layout and mappers.

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by ChanochShayner** [comment]
The current status -
We encountered a problem for incident fields that belong to paid packs and are in content.
We are currently working on a way to merge the `id_set` of the content and private repo, to resolve this issue.
@guykeller FYI

File: demisto_sdk/commands/common/hook_validations/mapper.py [hunk 1]
```diff
@@ -97,3 +101,32 @@ def is_valid_type(self):
             if self.handle_error(error_message, error_code, file_path=self.file_path):
                 return False
         return True
+
+    def is_incident_field_exist(self, id_set_file) -> bool:
+        if not id_set_file:
+            click.secho("Skipping mapper incident field validation. Could not read id_set.json.", fg="yellow")
+            return True
+
+        built_in_fields = [field.lower() for field in BUILT_IN_FIELDS] + LAYOUT_BUILT_IN_FIELDS
```
Comment (roysagi) [inline demisto_sdk/commands/common/hook_validations/mapper.py:110]: why LAYOUT_BUILD_IN_FIELDS in mapper validator?

**Review by roysagi — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/common/errors.py [hunk 2]
```diff
@@ -1085,6 +1087,11 @@ def invalid_file_path_layoutscontainer(file_name):
         return f'Invalid file name - {file_name}. layoutscontainer file name should start with ' \
                '"layoutscontainer-" prefix.'
 
+    @staticmethod
+    @error_code_decorator
+    def invalid_incident_field_in_layout(invalid_inc_fields_list):
+        return f'Your layout contains incident fields that do not exist in the content: {invalid_inc_fields_list}.'
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/errors.py:1103]: done

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 3]
```diff
@@ -118,3 +132,45 @@ def is_valid_type(self):
             if self.handle_error(error_message, error_code, file_path=self.file_path):
                 return False
         return True
+
+    def is_incident_field_exist(self, id_set_file) -> bool:
+        is_valid = True
+
+        if not id_set_file:
+            click.secho("Skipping mapper incident field validation. Could not read id_set.json.", fg="yellow")
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py:138]: done

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 4]
```diff
@@ -118,3 +132,45 @@ def is_valid_type(self):
             if self.handle_error(error_message, error_code, file_path=self.file_path):
                 return False
         return True
+
+    def is_incident_field_exist(self, id_set_file) -> bool:
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py:136]: done

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 5]
```diff
@@ -118,3 +132,45 @@ def is_valid_type(self):
...
+        mapper = self.current_file.get('mapping', {})
+
+        for key, value in mapper.items():
+            incident_fields = value.get('internalMapping', {})
+
+            for inc_name, inc_info in incident_fields.items():
+                mapper_incident_fields.append(inc_name)
+
+        content_incident_fields = []
+        content_all_incident_fields = id_set_file.get('IncidentFields')
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: done

File: demisto_sdk/commands/common/hook_validations/mapper.py (hunk 1 above)
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/mapper.py:110]: done

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 6]
```diff
@@ -118,3 +132,45 @@ def is_valid_type(self):
...
+            for _, inc_field in content_inc_field.items():
+                content_incident_fields.append(inc_field.get('name', ''))
+
+        content_indicator_fields = []
+        content_all_indicator_fields = id_set_file.get('IndicatorFields')
+        for content_ind_field in content_all_indicator_fields:
+            for _, ind_field in content_ind_field.items():
+                content_indicator_fields.append(ind_field.get('name', ''))
+
+        all_valid_fields = content_indicator_fields + content_incident_fields + LAYOUT_BUILT_IN_FIELDS + BUILT_IN_FIELDS
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: done

**Review by guykeller — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/common/errors.py [hunk 7]
```diff
@@ -1091,6 +1093,11 @@ def invalid_file_path_layoutscontainer(file_name):
...
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/errors.py]: This validation is not related to a legal field name it checks if it exists in the content

File: demisto_sdk/commands/common/errors.py [hunk 8]
```diff
@@ -1156,6 +1163,11 @@ def missing_from_version_in_mapper():
     def invalid_type_in_mapper():
         return 'Mappers type must be mapping-incoming or mapping-outgoing'
 
+    @staticmethod
+    @error_code_decorator
+    def invalid_incident_field_in_mapper(invalid_inc_fields_list):
+        return f'Your mapper contains incident fields that do not exist in the content: {invalid_inc_fields_list}.'
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/errors.py]: same

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 9]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
             if self.handle_error(error_message, error_code, file_path=self.file_path):
                 return False
         return True
+
+    def is_valid_incident_field(self) -> bool:
+        mapper_incident_fields = []
+
+        mapper = self.current_file.get('mapping', {})
+
+        for key, value in mapper.items():
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py:154]: done

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 10]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
...
+            incident_fields = value.get('internalMapping', {})
+
+            for inc_name, inc_info in incident_fields.items():
+                mapper_incident_fields.append(inc_name)
+
+        id_set_path = IDSetValidator.ID_SET_PATH
+        id_set = open_id_set_file(id_set_path)
+        if not id_set_path or not os.path.isfile(id_set_path):
+            id_set = IDSetCreator(print_logs=False).create_id_set()
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: already removed.

[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: already removed.

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 11]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
...
+
+        content_incident_fields = []
+        content_all_incident_fields = id_set.get('IncidentFields')
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: already removed.

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 12]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
...
+
+    def is_valid_incident_field(self) -> bool:
+        mapper_incident_fields = []
+
+        mapper = self.current_file.get('mapping', {})
+
+        for key, value in mapper.items():
+            incident_fields = value.get('internalMapping', {})
+
+            for inc_name, inc_info in incident_fields.items():
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py:157]: done

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 13]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
...
+        id_set_path = IDSetValidator.ID_SET_PATH
+        id_set = open_id_set_file(id_set_path)
+        if not id_set_path or not os.path.isfile(id_set_path):
+            id_set = IDSetCreator(print_logs=False).create_id_set()
+
+        content_incident_fields = []
+        content_all_incident_fields = id_set.get('IncidentFields')
+        for content_inc_field in content_all_incident_fields:
+            for _, inc_field in content_inc_field.items():
+                content_incident_fields.append(inc_field.get('name', ''))
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: already removed.

File: demisto_sdk/commands/common/hook_validations/layout.py [hunk 14]
```diff
@@ -26,7 +31,8 @@ def is_valid_layout(self, validate_rn=True) -> bool:
                     self.is_valid_from_version(),
                     self.is_valid_to_version(),
                     self.is_to_version_higher_than_from_version(),
-                    self.is_valid_file_path()
+                    self.is_valid_file_path(),
+                    # self.is_valid_incident_field()
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/layout.py]: Already removed

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 15]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
...
+
+        invalid_inc_fields_list = []
+        for mapper_inc_field in mapper_incident_fields:
+            if mapper_inc_field not in content_incident_fields and mapper_inc_field not in BUILT_IN_FIELDS \
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: Already removed

File: demisto_sdk/commands/common/hook_validations/classifier.py [hunk 16]
```diff
@@ -118,3 +135,39 @@ def is_valid_type(self):
...
+                    and mapper_inc_field not in LAYOUT_BUILT_IN_FIELDS:
+                invalid_inc_fields_list.append(mapper_inc_field) if mapper_inc_field not in invalid_inc_fields_list \
+                    else None
```
[the remark this answers is not in the document]
Comment (ChanochShayner) [reply demisto_sdk/commands/common/hook_validations/classifier.py]: Already removed

**Review by roysagi — APPROVED** [review]

**Review by guykeller — APPROVED** [review]

**Pull request merged**